core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala

```diff
@@ -1321,6 +1321,40 @@ class FsHistoryProviderSuite extends SparkFunSuite with Matchers with Logging {
     assertSerDe(serializer, attemptInfoWithIndex)
   }
 
+  test("SPARK-29043: clean up specified event log") {
+    def getLogPath(logFile: File): String = {
+      val uri = logFile.toURI
+      uri.getScheme + File.pathSeparator + uri.getPath
+    }
+
+    val clock = new ManualClock()
+    val conf = createTestConf().set(MAX_LOG_AGE_S.key, "2d")
+    val provider = new FsHistoryProvider(conf, clock)
+
+    // create an invalid application log file
+    val nonValidLogFile = newLogFile("NonValidLogFile", None, inProgress = true)
+    nonValidLogFile.createNewFile()
+    writeFile(nonValidLogFile, None,
+      SparkListenerApplicationStart(nonValidLogFile.getName, None, 1L, "test", None))
+    nonValidLogFile.setLastModified(clock.getTimeMillis())
+
+    // create a valid application log file
+    val validLogFile = newLogFile("validLogFile", None, inProgress = true)
+    validLogFile.createNewFile()
+    writeFile(validLogFile, None,
+      SparkListenerApplicationStart(validLogFile.getName, Some("local_123"), 1L, "test", None))
+    validLogFile.setLastModified(clock.getTimeMillis())
+
+    provider.checkForLogs()
+    clock.advance(TimeUnit.DAYS.toMillis(2))
+    provider.checkAndCleanLog(getLogPath(nonValidLogFile))
+    assert(new File(testDir.toURI).listFiles().size === 1)
```

Comment:
This isn't exactly the same thing the FsHistoryProvider class is doing.

That class is calling `checkAndCleanLog` as part of `checkForLogs`. The latter method updates things in the log metadata, like the `lastUpdated` field. So really what should be happening here is that the log file is deleted after you call `checkForLogs` (or more specifically, `doMergeApplicationListing`), not after you call `checkAndCleanLog` directly.

Reply:
Thanks, I have modified the UT.
I set `spark.history.fs.cleaner.maxAge` to 0 and create an invalid application log and a valid application log.
It seems that the checkAndCleanLog would clean this invalid application log after do merge.

And the cleanLogs method would clean the valid application log when it is expired.